Formal verification for the Governor system

Amxx opened the pull request:
WIP

frangio commented on certora/run.js:
```diff
@@ -8,26 +8,27 @@
 const events = require('events');
 const limit = require('p-limit')(MAX_PARALLEL);
 
+const strToRegex = str => new RegExp(`^${str.replace(/[.+?^${}()|[\]\\]/g, '\\$&').replace(/[*]/g, '.$&')}$`);
```
We can just use the `micromatch` dependency instead of this. It's already declared in our dev dependencies.

Amxx replied:
I'll check it out !

frangio commented on certora/specs.js:
```diff
@@ -0,0 +1,51 @@
+const product = (...arrays) => arrays.reduce((a, b) => a.flatMap(ai => b.map(bi => [ai, bi].flat())));
```
```suggestion
const product = (...arrays) => arrays.reduce((a, b) => a.flatMap(ai => b.map(bi => [ai, bi])));
```

frangio replied:
Ok I see it's for n > 2...

Amxx replied:
what you propose limit the number of array to 2. The current version goes beyond:

```javascript
> product([1,2],[true,false],['a','b'])
[
  [ 1, true, 'a' ],
  [ 1, true, 'b' ],
  [ 1, false, 'a' ],
  [ 1, false, 'b' ],
  [ 2, true, 'a' ],
  [ 2, true, 'b' ],
  [ 2, false, 'a' ],
  [ 2, false, 'b' ]
]
```

Amxx replied:
that can be usefull if you ever want to combine "specs" with "contract" with some link (like the token used)

frangio commented:
We need to merge `master` being careful to include the new specs, which at the moment seem to be `DoubleEndedQueue` and `ERC721`.

frangio commented on certora/specs/GovernorFunctions.spec:
```diff
@@ -0,0 +1,277 @@
+
+    address[] targets; uint256[] values; bytes[] calldatas; string descr;
+    require pId == hashProposal(targets, values, calldatas, descr);
+    //require validString(descr);
```
Can this line be removed?

Amxx replied:
the real answer is: I don't know, this rule timeout so I can't really say much about it.

frangio commented on certora/specs/GovernorFunctions.spec:
```diff
@@ -0,0 +1,277 @@
+
+    // no side-effect
+    assert state(e, otherId)         != otherStateBefore => otherId == pId;
+    assert proposalSnapshot(otherId) == otherVoteStart;
+    assert proposalDeadline(otherId) == otherVoteEnd;
+    assert proposalProposer(otherId) == otherProposer;
```
Slightly confused, why do these work even for `otherId == pId`?

Amxx replied:
In this rule we do `state(e, otherId)`. This means that `otherId` has a valid state, which means its been proposed (because state reverts if the proposal doesn't exist).

So we know `otherId` exists, and as been propose. 

`propose_liveness` is supposed (if it was working) to prove that `propose(...)` reverts if the proposal already exist. Since this rule does call propose without `@withrevert`, we know the propose worked, which implies `otherId == pId`.

So technically, we (expect to) know for sure that `otherId == pId`.

Amxx replied:
I think these (failing) rules should be rewritten using `safeState`

Amxx commented on certora/specs/Governor.helpers.spec:
```diff
@@ -0,0 +1,163 @@
+definition EXPIRED()   returns uint8 = 6;
+definition EXECUTED()  returns uint8 = 7;
+
+function safeState(env e, uint256 pId) returns uint8 {
```
added a comment

Amxx commented on certora/specs/Governor.helpers.spec:
```diff
@@ -0,0 +1,163 @@
+│ Filters                                                                                                             │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+definition skip(method f) returns bool =
```
Agree, but I don't have a better name on the top of my head.

This is a list of function that we don't consider when doing a "any function" rule. The point is to reduce the search space avoid some timeouts.

Example of usage:
```
rule immutableFieldsAfterProposalCreation(uint256 pId, env e, method f, calldataarg args)
    filtered { f -> !skip(f) }
{
...
}
```

Amxx commented on certora/specs/Governor.helpers.spec:
```diff
@@ -0,0 +1,163 @@
+    }
+}
+
+function helperFunctionsWithRevert(env e, method f, uint256 pId) {
```
added a long comment

Amxx commented on certora/specs/GovernorInvariants.spec:
```diff
@@ -0,0 +1,151 @@
+        getAbstainVotes(pId) > 0
+    ) => proposalSnapshot(pId) < clock(e)
+    {
+        preserved with (env e2) {
+            require clock(e) == clock(e2);
```
I'm not 100% sure, I'm starting to winder if it makes sens.

The issue is that all the envfree functions are not tied to a clock :/

Amxx commented on certora/specs/GovernorStates.spec:
```diff
@@ -0,0 +1,176 @@
+    uint8 stateAfter  = state(e, pId);
+
+    assert (stateBefore != stateAfter) => (
+        stateBefore == UNSET() => (
+            stateAfter == PENDING() && f.selector == propose(address[],uint256[],bytes[],string).selector
+        ) &&
+        stateBefore == PENDING() => (
+            (stateAfter == CANCELED() && f.selector == cancel(address[],uint256[],bytes[],bytes32).selector)
+        ) &&
+        stateBefore == SUCCEEDED() => (
+            (stateAfter == QUEUED()   && f.selector == queue(address[],uint256[],bytes[],bytes32).selector) ||
+            (stateAfter == EXECUTED() && f.selector == execute(address[],uint256[],bytes[],bytes32).selector)
+        ) &&
+        stateBefore == QUEUED() => (
+            (stateAfter == EXECUTED() && f.selector == execute(address[],uint256[],bytes[],bytes32).selector)
+        ) &&
+        stateBefore == ACTIVE()    => false &&
+        stateBefore == CANCELED()  => false &&
+        stateBefore == DEFEATED()  => false &&
+        stateBefore == EXECUTED()  => false
```
I'm not sure about the precedence of
```
A && B || C && D || E && F
```
So I added additional `(`, `)`

Amxx commented on certora/specs/GovernorStates.spec:
```diff
@@ -0,0 +1,176 @@
+    bool    voteSuccess   = voteSucceeded(pId);
+
+    // Pending: before vote starts
+    assert currentState == PENDING() => (
```
the goal here is to prove that some properties are true about each state. 
- If you are in state A, than X and Y
- If you are in state B, than Z 
- ...

There might indeed be cases where we can say
- If X and Y, than you are in state A
but where you cannot say 
- If Z, ... might be state B but might also be something else.

Should this rule remain consistent in how it treats all states, of should we prove more ?

frangio replied:
Let's keep the rule as it is now so we don't introduce new complexities at this point.

Amxx commented on certora/specs/GovernorStates.spec:
```diff
@@ -0,0 +1,176 @@
+    bool    voteSuccess   = voteSucceeded(pId);
+
+    // Pending: before vote starts
+    assert currentState == PENDING() => (
```
We can't say much about `quorumSuccess` because the quorum might be `= 0` (in which case no vote = success) or `> 0` (in which case no vote = no success)

For the `voteSuccess` its more complex. I order to say we don't have success, we need to "prove" that the votes counters are empty. This is done by `votesImplyCreated` with is part of `GovernorInvariants.spec`. If we want to prove it here, we need to import the invariant and require it. Note that this will cause all the runs of this spec file to re-verify the invariant.

frangio commented on certora/specs/GovernorStates.spec:
```diff
@@ -68,18 +57,26 @@ rule stateTransitionWait(uint256 pId, env e1, env e2) {
-        stateBefore == DEFEATED()  => false &&
-        stateBefore == EXECUTED()  => false
+        (stateBefore == PENDING() && stateAfter == ACTIVE()   ) ||
+        (stateBefore == PENDING() && stateAfter == DEFEATED() ) ||
+        (stateBefore == ACTIVE()  && stateAfter == SUCCEEDED()) ||
+        (stateBefore == ACTIVE()  && stateAfter == DEFEATED() ) ||
+        // Strange consequence of the timelock binding:
+        // When transitioning from ACTIVE to SUCCEEDED (because of the clock moving forward) the proposal state in
+        // the timelock is suddenly considered. Prior state set in the timelock can cause the proposal to already be
+        // queued, executed or canceled.
+        (stateBefore == ACTIVE()  && stateAfter == CANCELED()) ||
+        (stateBefore == ACTIVE()  && stateAfter == EXECUTED()) ||
+        (stateBefore == ACTIVE()  && stateAfter == QUEUED())
```
Not sure what we should do here. This rule is currently not looking good, as we would never expect a proposal to go (for example) from pending to defeated automatically, I think...

If the rule was working before (using `=>`) we should probably keep it like that.

Amxx replied:
I needed to add an additional "helper" invariant, but I managed to fix that part :)

Amxx commented:
Note: I wanted to do
```
node certora/run.js GovernorStates --options="--rule stateTransitionWait"
```
But that caused issues with options including spaces. So I moved the `.flatMap(opt => opt.split(' '))` to inside the `runCertora` function

@frangio can you see any issue with that ?

frangio commented:
No, that makes sense. I don't think it's ideal the way we're handling options but it works.

Amxx commented:
FYI: main (only?) remaining blocker is the `stateTransitionFn` function timming out.

It appears that it only reason it used to pass was because of poorly written formula and mistakes in the associativity of operators. Rewritting them cleanly (or adding parenthesis) causes it to timeout. An option could be to split that rule into multiple rules with limited logic in each one.
